### PR TITLE
[GLUTEN-1577][CORE] Respect spark's config for case sensitive when get attribute name

### PR DESCRIPTION
**Pull request by exmy**
## What changes were proposed in this pull request?

(Fixes: \#1577)

## How was this patch tested?
ut

### CODE REVIEW

**Comment by FelixYBW** [comment]
@PHILO-HE is the column name case sensitive issue fully solved?

**Comment by exmy** [comment]
@rui-mo @zzcclp Could you review this pr? Thanks!

**Comment by PHILO-HE** [comment]
> @PHILO-HE is the column name case sensitive issue fully solved?

We already have some code logic for reading parquet column in case-insensitive mode.
@exmy, could you provide some case for producing the issue?

**Comment by exmy** [comment]
@PHILO-HE I encountered the issue with the ORC column, specifically when constructing a NamedStruct in the `StorageJoinBuilder` class exist in `backends-clickhouse` module. I'm not sure if Velox has this problem.
> We already have some code logic for reading parquet column in case-insensitive mode.

Could you point out where this piece of code logic is located?

**Comment by PHILO-HE** [comment]
> @PHILO-HE I encountered the issue with the ORC column, specifically when constructing a NamedStruct in the `StorageJoinBuilder` class exist in `backends-clickhouse` module. I'm not sure if Velox has this problem.
> 
> > We already have some code logic for reading parquet column in case-insensitive mode.
> 
> Could you point out where this piece of code logic is located?

@exmy, a method called `normalizeColName` in `BasicScanExecTransformer` will do the case conversion. And there are also some fixes for velox in native code.

**Comment by exmy** [comment]
> @exmy, a method called `normalizeColName` in `BasicScanExecTransformer` will do the case conversion. It is applicable for parquet/ORC. And there are also some fixes for velox in native code.

@PHILO-HE Thanks! I noticed that there may be other places where the case conversion also needs to be applied, not just in `BasicScanExecTransformer`, for example, `PlanNodesUtil.genProjectionsPlanNode`.

We may have the plan:
```
 CHBroadcastHashJoinExecTransformer [scene#2, experiment_id#3, layer#4], [scene#8, experiment_id#9, layer#10], Inner, BuildLeft, false
      :     :- ColumnarBroadcastExchange HashedRelationBroadcastMode(List(input[0, string, true], input[1, string, true], input[2, string, true]),false), [plan_id=112]
      :     :  +- ^(1) ProjectExecTransformer [scene#2, experiment_id#3, LAYER#4]
      :     :     +- ^(1) FilterExecTransformer ((isnotnull(scene#2) AND isnotnull(experiment_id#3)) AND isnotnull(layer#4))
      :     :        +- ^(1) NativeFileScan orc tbl[scene#2,experiment_id#3,LAYER#4,DAY#6,hour#7] Batched: true, DataFilters: [isnotnull(scene#2), isnotnull(experiment_id#3), isnotnull(LAYER#4)], Format: ORC, Location: InMemoryFileIndex(1 paths)[..., PartitionFilters: [isnotnull(DAY#6), isnotnull(hour#7), (DAY#6 = 2023-03-15), (hour#7 = 20)], PushedFilters: [IsNotNull(scene), IsNotNull(experiment_id), IsNotNull(LAYER)], ReadSchema: struct<scene:string,experiment_id:string,LAYER:string>
```
ProjectExecTransformer have a  output attribute called `LAYER`, which may also need do the case conversion.

File: gluten-core/src/main/scala/io/glutenproject/expression/ConverterUtils.scala [hunk 1]
```diff
@@ -76,11 +79,16 @@ object ConverterUtils extends Logging {
   }
 
   def getShortAttributeName(attr: Attribute): String = {
```
[the remark this answers is not in the document]
Comment (exmy) [reply gluten-core/src/main/scala/io/glutenproject/expression/ConverterUtils.scala:81]: This function is called by `converter.genColumnNameWithExprId` which is used in many other places, and there may be other places where the case conversion also needs to be applied. What do you think？

[the remark this answers is not in the document]
Comment (rui-mo) [reply gluten-core/src/main/scala/io/glutenproject/expression/ConverterUtils.scala:81]: Got it. Thanks for checking. Let's keep it there then.

**Comment by PHILO-HE** [comment]
> We may have the plan:
> 
> ```
>  CHBroadcastHashJoinExecTransformer [scene#2, experiment_id#3, layer#4], [scene#8, experiment_id#9, layer#10], Inner, BuildLeft, false
>       :     :- ColumnarBroadcastExchange HashedRelationBroadcastMode(List(input[0, string, true], input[1, string, true], input[2, string, true]),false), [plan_id=112]
>       :     :  +- ^(1) ProjectExecTransformer [scene#2, experiment_id#3, LAYER#4]
>       :     :     +- ^(1) FilterExecTransformer ((isnotnull(scene#2) AND isnotnull(experiment_id#3)) AND isnotnull(layer#4))
>       :     :        +- ^(1) NativeFileScan orc tbl[scene#2,experiment_id#3,LAYER#4,DAY#6,hour#7] Batched: true, DataFilters: [isnotnull(scene#2), isnotnull(experiment_id#3), isnotnull(LAYER#4)], Format: ORC, Location: InMemoryFileIndex(1 paths)[..., PartitionFilters: [isnotnull(DAY#6), isnotnull(hour#7), (DAY#6 = 2023-03-15), (hour#7 = 20)], PushedFilters: [IsNotNull(scene), IsNotNull(experiment_id), IsNotNull(LAYER)], ReadSchema: struct<scene:string,experiment_id:string,LAYER:string>
> ```
> 
> ProjectExecTransformer have a output attribute called `LAYER`, which may also need do the case conversion.

Thanks for further clarification! For velox backend, I think there is a mechanism to bind column index. I'm not sure whether there is still any potential issue.

File: gluten-ut/spark32/src/test/scala/org/apache/spark/sql/GlutenJoinSuite.scala [hunk 2]
```diff
@@ -40,4 +40,18 @@ class GlutenJoinSuite extends JoinSuite with GlutenSQLTestsTrait {
     // NaN is not supported currently, just skip.
     "NaN and -0.0 in join keys"
   )
+
+  test("test case sensitive for BHJ") {
```
Comment (PHILO-HE) [inline gluten-ut/spark32/src/test/scala/org/apache/spark/sql/GlutenJoinSuite.scala:44]: Please add `GLUTEN_TEST` in the title.

**Comment by PHILO-HE** [comment]
In the long time, maybe it is better to have a mechanism of binding index or exprID to refer a column, instead of through column name. Anyway, we can get it merged firstly. Thanks!

**Review by PHILO-HE — DISMISSED** [review]

File: gluten-ut/spark32/src/test/scala/org/apache/spark/sql/GlutenJoinSuite.scala (hunk 2 above)
Comment (exmy) [reply gluten-ut/spark32/src/test/scala/org/apache/spark/sql/GlutenJoinSuite.scala]: Done

**Review by FelixYBW — APPROVED** [review]

**Pull request merged**